### PR TITLE
Fixed #29728 -- Prevented session resaving if CSRF cookie is unchanged.

### PR DESCRIPTION
**Pull request by nijel**
https://code.djangoproject.com/ticket/29728

### CODE REVIEW

**Comment by charettes** [comment]
Hello @nijel, thanks for the patch!

Could you add a test to [`CsrfViewMiddlewareUseSessionsTests` as well](https://github.com/django/django/blob/ee52044a278885bd9455dd59b1e16c5d5e2d68ce/tests/csrf_tests/tests.py#L667)?

You should be able to assert against `request.session.modified`.

**Comment by nijel** [comment]
I've tried, but looks too complex for me right now. The problem being that all tests currently use TestingHttpRequest, which does not have real session storage, but just a dict attached as a session...

**Comment by nijel** [comment]
Okay, I've replaced the session for tests, hopefully that's correct way to do that.

File: tests/csrf_tests/tests.py [hunk 1]
```diff
@@ -693,6 +693,19 @@ def test_process_response_get_token_used(self):
         ensure_csrf_cookie_view(req)
         self.assertTrue(req.session.get(CSRF_SESSION_KEY, False))
 
+    def test_session_modify(self):
+        """The session should not be modified if not necessary."""
+        req = self._get_GET_no_csrf_cookie_request()
+        self.mw.process_view(req, ensure_csrf_cookie_view, (), {})
+        resp = ensure_csrf_cookie_view(req)
+        self.mw.process_response(req, resp)
+        self.assertTrue(req.session.get(CSRF_SESSION_KEY, False))
```
Comment (charettes) [inline tests/csrf_tests/tests.py]: I think `assertIsNotNone(req.session.get(CSRF_SESSION_KEY))` would better express the existence assertion here.

File: tests/csrf_tests/tests.py [hunk 2]
```diff
@@ -693,6 +693,19 @@ def test_process_response_get_token_used(self):
         ensure_csrf_cookie_view(req)
         self.assertTrue(req.session.get(CSRF_SESSION_KEY, False))
 
+    def test_session_modify(self):
+        """The session should not be modified if not necessary."""
```
Comment (charettes) [inline tests/csrf_tests/tests.py]: Avoid the conditionals woridng-- _The session is not modified if not necessary._

File: tests/csrf_tests/tests.py (hunk 1 above)
Comment (nijel) [reply tests/csrf_tests/tests.py]: Well I've shamelessly copied the code in above function, but you're right.

**Pull request merged**